src/cart/crt_context.c

```diff
@@ -948,6 +948,17 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
 	D_INIT_LIST_HEAD(&submit_list);
 
 	D_MUTEX_LOCK(&epi->epi_mutex);
+
+	/* Prevent simultaneous untrack from progress thread and
+	 * main rpc execution thread.
+	 */
+	if (rpc_priv->crp_in_binheap == 0) {
+		RPC_TRACE(DB_NET, rpc_priv,
+			"rpc is not tracked already.\n");
+		D_MUTEX_UNLOCK(&epi->epi_mutex);
+		return;
```

Comment:
This mean you will skip the process waitq. Is it intentionally?

Reply:
yes, ideally this function should not be called twice at all for the same rpc, but in the corner case it can happen (when timeout happens relatively closely to actual hg_forward() failing).